### PR TITLE
Issue #150 Bug Fix and unittest

### PR DESCRIPTION
**Pull request by armartirosyan**
The code fixes bug reported under issue #150  and contains supporting unittest code.

### CODE REVIEW

File: test/unit/test_devices/test_ios_device.py [hunk 1]
```diff
@@ -341,6 +341,45 @@ def test_enable_from_config(self):
         self.device.native.check_config_mode.assert_called()
         self.device.native.exit_config_mode.assert_called()
 
+    @mock.patch.object(IOSDevice, "_image_booted", side_effect=[False, True])
+    @mock.patch.object(IOSDevice, "set_boot_options")
+    @mock.patch.object(IOSDevice, "reboot")
+    @mock.patch.object(IOSDevice, "_wait_for_device_reboot")
+    def test_install_os_image_booted_false_true(self, mock_wait, mock_reboot, mock_set_boot, mock_image_booted):
```
Comment (jmcgill298) [inline test/unit/test_devices/test_ios_device.py:348]:
```suggestion
    def test_install_os(self, mock_wait, mock_reboot, mock_set_boot, mock_image_booted):
```

File: test/unit/test_devices/test_ios_device.py [hunk 2]
```diff
@@ -341,6 +341,45 @@ def test_enable_from_config(self):
...
+        mock_set_boot.assert_called()
+        mock_reboot.assert_called()
+        mock_wait.assert_called()
+        self.assertEqual(state, True)
+
+    @mock.patch.object(IOSDevice, "_image_booted", side_effect=[True, True])
+    @mock.patch.object(IOSDevice, "set_boot_options")
+    @mock.patch.object(IOSDevice, "reboot")
+    @mock.patch.object(IOSDevice, "_wait_for_device_reboot")
+    def test_install_os_image_booted_true_true(self, mock_wait, mock_reboot, mock_set_boot, mock_image_booted):
```
Comment (jmcgill298) [inline test/unit/test_devices/test_ios_device.py:359]:
```suggestion
    def test_install_os_already_installed(self, mock_wait, mock_reboot, mock_set_boot, mock_image_booted):
```

File: test/unit/test_devices/test_ios_device.py [hunk 3]
```diff
@@ -341,6 +341,45 @@ def test_enable_from_config(self):
...
+        mock_set_boot.assert_not_called()
+        mock_reboot.assert_not_called()
+        mock_wait.assert_not_called()
+        self.assertEqual(state, False)
+
+    @mock.patch.object(IOSDevice, "_image_booted", side_effect=[False, False])
+    @mock.patch.object(IOSDevice, "set_boot_options")
+    @mock.patch.object(IOSDevice, "reboot")
+    @mock.patch.object(IOSDevice, "_wait_for_device_reboot")
+    def test_install_os_image_booted_false_false(self, mock_wait, mock_reboot, mock_set_boot, mock_image_booted):
```
Comment (jmcgill298) [inline test/unit/test_devices/test_ios_device.py:370]:
```suggestion
    def test_install_os_error(self, mock_wait, mock_reboot, mock_set_boot, mock_image_booted):
```

File: test/unit/test_devices/test_ios_device.py [hunk 4]
```diff
@@ -341,6 +341,45 @@ def test_enable_from_config(self):
         self.device.native.check_config_mode.assert_called()
         self.device.native.exit_config_mode.assert_called()
 
+    @mock.patch.object(IOSDevice, "_image_booted", side_effect=[False, True])
+    @mock.patch.object(IOSDevice, "set_boot_options")
+    @mock.patch.object(IOSDevice, "reboot")
+    @mock.patch.object(IOSDevice, "_wait_for_device_reboot")
+    def test_install_os_image_booted_false_true(self, mock_wait, mock_reboot, mock_set_boot, mock_image_booted):
+        state=self.device.install_os(BOOT_IMAGE)
```
Comment (jmcgill298) [inline test/unit/test_devices/test_ios_device.py:349]:
```suggestion
        state = self.device.install_os(BOOT_IMAGE)
```

File: test/unit/test_devices/test_ios_device.py [hunk 5]
```diff
@@ -341,6 +341,45 @@ def test_enable_from_config(self):
...
+        mock_reboot.assert_called()
+        mock_wait.assert_called()
+        self.assertEqual(state, True)
+
+    @mock.patch.object(IOSDevice, "_image_booted", side_effect=[True, True])
+    @mock.patch.object(IOSDevice, "set_boot_options")
+    @mock.patch.object(IOSDevice, "reboot")
+    @mock.patch.object(IOSDevice, "_wait_for_device_reboot")
+    def test_install_os_image_booted_true_true(self, mock_wait, mock_reboot, mock_set_boot, mock_image_booted):
+        state=self.device.install_os(BOOT_IMAGE)
```
Comment (jmcgill298) [inline test/unit/test_devices/test_ios_device.py:360]:
```suggestion
        state = self.device.install_os(BOOT_IMAGE)
```

File: test/unit/test_devices/test_ios_device.py [hunk 6]
```diff
@@ -341,6 +341,45 @@ def test_enable_from_config(self):
...
+    @mock.patch.object(IOSDevice, "reboot")
+    @mock.patch.object(IOSDevice, "_wait_for_device_reboot")
+    def test_install_os_image_booted_false_true(self, mock_wait, mock_reboot, mock_set_boot, mock_image_booted):
+        state=self.device.install_os(BOOT_IMAGE)
+        mock_set_boot.assert_called()
+        mock_reboot.assert_called()
+        mock_wait.assert_called()
+        self.assertEqual(state, True)
+
+    @mock.patch.object(IOSDevice, "_image_booted", side_effect=[True, True])
```
Comment (jmcgill298) [inline test/unit/test_devices/test_ios_device.py:355]:
```suggestion
    @mock.patch.object(IOSDevice, "_image_booted", side_effect=[True])
```

File: test/unit/test_devices/test_ios_device.py [hunk 7]
```diff
@@ -341,6 +341,45 @@ def test_enable_from_config(self):
...
+    @mock.patch.object(IOSDevice, "set_boot_options")
+    @mock.patch.object(IOSDevice, "reboot")
+    @mock.patch.object(IOSDevice, "_wait_for_device_reboot")
+    def test_install_os_image_booted_true_true(self, mock_wait, mock_reboot, mock_set_boot, mock_image_booted):
+        state=self.device.install_os(BOOT_IMAGE)
+        mock_set_boot.assert_not_called()
+        mock_reboot.assert_not_called()
+        mock_wait.assert_not_called()
+        self.assertEqual(state, False)
```
Comment (jmcgill298) [inline test/unit/test_devices/test_ios_device.py:364]:
```suggestion
        self.assertEqual(state, False)
        mock_image_booted.assert_called_once()
```

File: test/unit/test_devices/test_ios_device.py [hunk 8]
```diff
@@ -341,6 +341,45 @@ def test_enable_from_config(self):
...
+    @mock.patch.object(IOSDevice, "_image_booted", side_effect=[True, False])
+    @mock.patch.object(IOSDevice, "set_boot_options")
+    @mock.patch.object(IOSDevice, "reboot")
+    @mock.patch.object(IOSDevice, "_wait_for_device_reboot")
+    def test_install_os_image_booted_true_false(self, mock_wait, mock_reboot, mock_set_boot, mock_image_booted):
+        state=self.device.install_os(BOOT_IMAGE)
+        mock_set_boot.assert_not_called()
+        mock_reboot.assert_not_called()
+        mock_wait.assert_not_called()
+        self.assertEqual(state, False)
```
Comment (jmcgill298) [inline test/unit/test_devices/test_ios_device.py:382]: This is the same test as `test_os_install_already_installed` and should be removed

**Review by jmcgill298 — COMMENTED** [review]
Thanks @armartirosyan 

Overall, this looks good, a few minor changes are needed, and also make sure that black and flake8 tests pass. You can run `tox -e black, flake8` locally to test before pushing back up.

File: test/unit/test_devices/test_ios_device.py (hunk 1 above)
Comment (armartirosyan) [reply test/unit/test_devices/test_ios_device.py]: Done

File: test/unit/test_devices/test_ios_device.py (hunk 2 above)
Comment (armartirosyan) [reply test/unit/test_devices/test_ios_device.py]: Fixed

File: test/unit/test_devices/test_ios_device.py (hunk 4 above)
Comment (armartirosyan) [reply test/unit/test_devices/test_ios_device.py]: Fixed

File: test/unit/test_devices/test_ios_device.py (hunk 5 above)
Comment (armartirosyan) [reply test/unit/test_devices/test_ios_device.py]: Fixed

File: test/unit/test_devices/test_ios_device.py (hunk 6 above)
Comment (armartirosyan) [reply test/unit/test_devices/test_ios_device.py]: Fixed

File: test/unit/test_devices/test_ios_device.py (hunk 7 above)
Comment (armartirosyan) [reply test/unit/test_devices/test_ios_device.py:365]: Fixed

File: test/unit/test_devices/test_ios_device.py (hunk 8 above)
Comment (armartirosyan) [reply test/unit/test_devices/test_ios_device.py]: Fixed

**Comment by jmcgill298** [comment]
Thanks @armartirosyan

**Pull request merged**